File: 456. 132 Pattern/find132pattern.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
const find132pattern = (nums) => {
  const stack = [];
  let value3InPattern = -Infinity;

  for (let i = nums.length - 1; i >= 0; i--) {
    if (nums[i] < value3InPattern) {
      return true;
    } else {
      while (stack.length && stack[stack.length - 1] < nums[i]) {
        value3InPattern = stack.pop()
      }
    }

    stack.push(nums[i])
  }

  return false;
};

const nums = [-1, 3, 2, 0];
console.log('find132pattern: ', find132pattern(nums));

File: 456. 132 Pattern/find132pattern.py
from typing import List

class Solution:
  def find132pattern(self, nums: List[int]) -> bool:
    stack = []
    minSoFar = nums[0]

    for n in nums[1:]:
      while stack and n >= stack[-1][0]:
        stack.pop()

      if stack and n > stack[-1][1]:
        return True
      
      stack.append([n, minSoFar])
      minSoFar = min(minSoFar, n)
    
    return False

nums = [3, 1, 4, 2]
solution = Solution()
print('find132pattern: ', solution.find132pattern(nums))
